Implementation of Boundary Intersection Check.

Amon-Cyn opened the pull request:
### Description:

New check named BoundaryIntersectionCheck added.
The purpose of this check is to verify that way belonging to the boundary relation is not intersecting with way belonging to another boundary relation. Such intersection is allowed only if boundaries belongs to different types (different boundary=* tag).

### Potential Impact:

None, just addition of a new check.

### Unit Test Approach:

Unit tests added for checking various boundary intersections and touches.

### Test Results:

As a result of a check boundary intersecting 
387 flags were reviewed in 5 countries. The table below provides details about the regions and flags analyzed. 94% of flags were found to be True Positives. 6% of flags were False Positives.

  |   | Flag Count | Valid Use Cases | Invalid Use Cases |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | --
Country | ISO | Total Flags | Sampled | Sampling % | Use Case 1 | Use Case 1 | TP | TP % | FP | FP%
Czech Republic | CZE | 157 | 37 | 24% | 37 | 0 | 35 | 95% | 2 | 5%
Uganda | UGA | 91 | 34 | 37% | 34 | 0 | 34 | 100% | 0 | 0%
Brazil | BRA | 200 | 88 | 44% | 88 | 0 | 80 | 91% | 8 | 9%
Portugal | PRT | 446 | 95 | 21% | 95 | 0 | 92 | 97% | 3 | 3%
Australia | AUS | 768 | 133 | 17% | 133 | 0 | 124 | 93% | 9 | 7%
TOTAL: | 894 | 387 | 43% | 387 | 0 | 365 | 94% | 22 | 6%

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+import org.openstreetmap.atlas.tags.annotations.validation.Validators;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+
+/**
```
Please add a description of the check here.

Bentleysb reviewed (commented):
This is a very interesting check. I am excited to see it. it looks like there are a bunch of methods that can be combined. This could also use more documentation. Please add an md file for the check in https://github.com/osmlab/atlas-checks/tree/dev/docs/checks. Please also add java docs to any method that is more than a couple line long. And also please add unit tests for the helper classes you created.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+        final Set<String> matchedTags = new HashSet<>();
+        for (final BoundaryPart currentBoundaryPart : relationBoundary.getBoundaryParts())
+        {
+            final Iterable<LineItem> lineItemsIntersecting = object.getAtlas()
```
If the atlas (`object.getAtlas()`) is saved beforehand, it need not be fetched for each BoundaryPart.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+                    .lineItemsIntersecting(currentBoundaryPart.getBounds(),
+                            this.getPredicateForLineItemsSelection(currentBoundaryPart,
+                                    relationBoundary.getTagToRelation().keySet()));
+            final Iterable<Area> areasIntersecting = object.getAtlas().areasIntersecting(
```
Same comment as before.

sayas01 reviewed (commented):
Thanks @Amon-Cyn for adding this! I would like to see more java docs for methods and classes added in this PR. Could you add some examples of the FPs in the PR description? Looks like there are 2 code smells. Please fix that as well. Thank you!

ameliaewatts commented:
Hey, @Amon-Cyn! Where do we stand with addressing these comments?

Amon-Cyn commented:
> Hey, @Amon-Cyn! Where do we stand with addressing these comments?

Hi, recently I am running on a really busy schedule. I hope that this week I will be able to start resolving issues mentioned in the comments. Sorry for the delay.

ameliaewatts commented:
> > Hey, @Amon-Cyn! Where do we stand with addressing these comments?
> 
> Hi, recently I am running on a really busy schedule. I hope that this week I will be able to start resolving issues mentioned in the comments. Sorry for the delay.

I think that's fine! @Bentleysb Is this okay for the next release then?

Amon-Cyn commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+    private Map<String, Relation> getRelationMap(final AtlasObject object)
+    {
+        final Map<String, Relation> tagToRelation = new HashMap<>();
+        if (object instanceof MultiRelation)
```
I am checking underlying relations for their tags if I am handling MultiRelation.

Amon-Cyn commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+        {
+            if (this.checkAreaAsBoundary(areaToCheck, boundaryTags))
+            {
+                return this.isCrossingNotTouching(boundaryPart.getWktGeometry(),
```
Yes, I am going through wkt to work on geometries. I browsed the package you mentioned, but I couldn't find any method that would take AtlasEntity as input and return JTS geometry as output. I found only two ways of extracting geometry from AtlasEntity, wkt and geojson.

Amon-Cyn commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+        });
+    }
+
+    private void processLineItems(final RelationBoundary relationBoundary,
```
This cannot be combined, although I have cleaned it up with extraction of common parts.

Amon-Cyn commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+
+    private List<LineItem> getLineItems(final LineItem lineItem)
+    {
+        if (lineItem instanceof MultiLine)
```
So i can process every LineItem that is part of MultiLine.

Bentleysb replied:
I think there is some confusion here as to what a `MultiLine` or `MultiRelation` is in `Atlas`. These are not entities with an extra dimension of geometry. They are the types of entities stored in a `MultiAtlas`. For the most part they are functionally identical to their sibling types for in a `PackedAtlas` or `DynamicAtlas` (`PackedLine`, `DynamicRelation`, etc). These different types of `Atlas` implementations are used to load the `Atlas` in different ways, but should generally be treated the same once loaded. So for the different entity types you should just use the parent classes of `Area`, `Edge`, `Line`, `Node`, `Point`, and `Relation`. 
It would be extremely rare that you need to check for specific Packed, Multi, or Dynamic entity in an Atlas Check. In this case your check will only work if it is run on a `MultiAtlas`. This is not the default for the legacy or sharded checks spark jobs.

Amon-Cyn replied:
@Bentleysb Thanks for the clarification, there was a confusion here. I will remove the obsolete code.

Amon-Cyn commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,403 @@
+    private Set<Relation> getBoundaries(final LineItem currentLineItem)
+    {
+        final Set<Relation> relations = currentLineItem.relations().stream()
+                .filter(relation -> relation instanceof MultiRelation).map(AtlasEntity::relations)
```
So i can process every Relation that is part of MultiRelation.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/utility/RelationIntersections.java:
```diff
@@ -0,0 +1,37 @@
+import org.openstreetmap.atlas.geography.atlas.items.LineItem;
+import org.openstreetmap.atlas.geography.atlas.items.Relation;
+
+/**
```
+1

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/utility/RelationIntersections.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class RelationIntersections
+{
+
```
nit: extra line

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/utility/RelationIntersections.java:
```diff
@@ -0,0 +1,37 @@
+    {
+        return this.intersections.keySet();
+    }
+
```
nit: extra line

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,420 @@
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+
+/**
+ * @author srachanski This check analyses whether there are some boundaries with the same tag that
```
Please add @author in a new line after the decsription

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,420 @@
+ */
+public class BoundaryIntersectionCheck extends BaseCheck<Long>
+{
+
```
nit: extra line

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java:
```diff
@@ -0,0 +1,420 @@
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    private boolean checkAtlasEntityAsBoundary(final AtlasEntity atlasEntity,
```
It would be great if you could add java doc to each of the methods added in the class with a description of what each method is doing.

sayas01 reviewed (commented):
@Amon-Cyn It would be great if you could add a java doc comment before each of the methods in the BoundaryIntersectionCheck class containing the method description.

ameliaewatts commented:
@Amon-Cyn Do you think we'll be able to get this out for tomorrow's release?

Amon-Cyn commented:
> @Amon-Cyn Do you think we'll be able to get this out for tomorrow's release?

@ameliaewatts I just commited all the requested changes.

sayas01 reviewed (approved):
Thanks for addressing the comments! Looks good!

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/validation/intersections/RelationBoundary.java:
```diff
@@ -0,0 +1,50 @@
+ *
+ * @author srachanski
+ */
+public class RelationBoundary
```
This doesn't need to happen now, but this would be a great candidate for a ComplexEntity in Atlas.
